core/src/mindustry/type/ammo/LiquidAmmoType.java

```diff
@@ -0,0 +1,56 @@
+package mindustry.type.ammo;
+
+import arc.graphics.*;
+import mindustry.content.*;
+import mindustry.entities.*;
+import mindustry.gen.*;
+import mindustry.graphics.*;
+import mindustry.type.*;
+
+public class LiquidAmmoType implements AmmoType{
+    public float range = 85f;
+    public int ammoPerLiquid = 15;
+    public Liquid liquid;
+    
+    public LiquidAmmoType(Liquid liquid){
+        this.liquid = liquid;
+    }
+    
+    public LiquidAmmoType(Liquid liquid, int ammoPerLiquid){
+        this.liquid = liquid;
+        this.ammoPerLiquid = ammoPerLiquid;
+    }
+    
+    public LiquidAmmoType(){
+    }
+    
+    @Override
+    public String icon(){
+        return liquid.emoji();
+    }
+    
+    @Override
+    public Color color(){
+        return liquid.color;
+    }
+    
+    @Override
+    public Color barColor(){
+        return Pal.ammo;
+    }
+    
+    @Override
+    public void resupply(Unit unit){
+        if(unit.type.ammoCapacity - unit.ammo < ammoPerLiquid) return;
+        
+        float range = unit.hitSize + this.range;
+        
+        Building build = Units.closestBuilding(unit.team, unit.x, unit.y, range, u -> u.canResupply() && u.liquids != null && (u.liquids.get(liquid) > 0));
+        
+        if(build != null){
+            Fx.itemTransfer.at(build.x, build.y, ammoPerLiquid / 2, liquid.color, unit);
+            unit.ammo = Math.min(unit.ammo + ammoPerLiquid, unit.type.ammoCapacity);
+            build.liquids.remove(liquid, 1f);
```

Comment:
This is wrong; you're testing if a block has >0 liquid, but removing 1 unit, which can lead to negative values. Liquid is not discrete like items are, you need to calculate how much ammo it will actually get, and withdraw in greater increments (1 liquid unit is practically nothing, and 15 ammo per liquid unit is ridiculous)

Reply:
(for reference, each second 3 units of cryofluid are pumped trough the most basic pump) 15 liquids for 1 ammo would me more fair? probably not but eh